src/opm/parser/eclipse/EclipseState/Grid/EclipseGrid.cpp

```diff
@@ -331,18 +553,441 @@ namespace Opm {
+
+        m_coord = makeCoordDxDyDzTops(dims, DX, DY, DZ, TOPS);
+        m_zcorn = makeZcornDzTops(dims, DZ, TOPS);
+
+        ZcornMapper mapper( getNX(), getNY(), getNZ());
+        zcorn_fixed = mapper.fixupZCORN( m_zcorn );
+    }
+
+    void EclipseGrid::calculateGeometryData() {
+
+        const std::array<int, 3> dims = getNXYZ();
+        int nCells = getCartesianSize();
+
+        std::vector<double> X(8, 0.0);
+        std::vector<double> Y(8, 0.0);
+        std::vector<double> Z(8, 0.0);
+
+        m_volume.clear();
+        m_cellCenter.clear();
+        m_dx.clear();
+        m_dy.clear();
+        m_dz.clear();
+        m_depth.clear();
+
+        
+        m_volume.resize(nCells);
+        m_cellCenter.resize(nCells);
+        m_dx.resize(nCells);
+        m_dy.resize(nCells);
+        m_dz.resize(nCells);
+        m_depth.resize(nCells);
+
+        std::array<int, 3> ijk;
+        size_t n = 0;
+        
+        for (size_t k = 0; k< getNZ(); k++) {
+            for (size_t j = 0; j< getNY(); j++) {
+                for (size_t i = 0; i< getNX(); i++) {
+
+                    ijk[0] = i;
+                    ijk[1] = j;
+                    ijk[2] = k;
+
+                    getCellCorners(ijk, dims, X, Y, Z );
+
+                    m_volume[n] = calculateCellVol(X, Y, Z);
+
+                    std::array<double, 3> center;
+
+                    center[0]=std::accumulate(X.begin(), X.end(), 0.0) / 8.0;
+                    center[1]=std::accumulate(Y.begin(), Y.end(), 0.0) / 8.0;
+                    center[2]=std::accumulate(Z.begin(), Z.end(), 0.0) / 8.0;
+
+                    m_cellCenter[n] = center;
+
+                    // calculate dx
+
+                    double x2 = (X[1]+X[3]+X[5]+X[7])/4.0;
+                    double y2 = (Y[1]+Y[3]+Y[5]+Y[7])/4.0;
+
+                    double x1 = (X[0]+X[2]+X[4]+X[6])/4.0;
+                    double y1 = (Y[0]+Y[2]+Y[4]+Y[6])/4.0;
+
+                    double dx = sqrt(pow((x2-x1), 2.0) + pow((y2-y1), 2.0) );
+
+                    // calculate dy
+
+                    x2 = (X[2]+X[3]+X[6]+X[7])/4.0;
+                    y2 = (Y[2]+Y[3]+Y[6]+Y[7])/4.0;
+
+                    x1 = (X[0]+X[1]+X[4]+X[5])/4.0;
+                    y1 = (Y[0]+Y[1]+Y[4]+Y[5])/4.0;
+
+                    double dy = sqrt(pow((x2-x1), 2.0) + pow((y2-y1), 2.0));
+
+                    // calculate dz
+
+                    double z2 = (Z[4]+Z[5]+Z[6]+Z[7])/4.0;
+                    double z1 = (Z[0]+Z[1]+Z[2]+Z[3])/4.0;
+
+                    double dz = z2-z1;
+
+                    m_dx[n] = dx;
+                    m_dy[n] = dy;
+                    m_dz[n] = dz;
+
+                    m_depth[n] =  (z2 + z1) / 2.0;
+
+                    n++;
+                }
+            }
+        }
+    }
+
+    void EclipseGrid::getCellCorners(const std::array<int, 3>& ijk, const std::array<int, 3>& dims,
+                                     std::vector<double>& X,
```

Comment:
If you are afraid of the size != 8 you could use `std::array<double,8>`?

Reply:
std::array<double,8> should have been the chosen from the beginning. Used vectors since calculateCellVol had vectors. I have now changed to std::array<double,8>. I have also updated the calculateCellVol function.

Is this the correct way to initialize an array with zero values ? 

```C++
std::array<double,8> X;
std::array<double,8> Y;
std::array<double,8> Z;

X.fill(0.0); 
Y.fill(0.0); 
Z.fill(0.0); 
```

or is there a more elegant way of doing this ?

Reply:
Is all checking for size redundant when using an array of fixed length like this ?

Reply:
```
std::array<double,8> X = {0};
```

Reply:
and yes, no need checking size, it's always 8.

Reply:
In general you may want to use exact type match i.e. `{0.0}` rather than `{0}`, due to some particulars of std::initializer_list (not strictly necessary here though).

Brief digression: I wish C++ would not accept the above `= {0}` but instead accept `= {0.0, ...}`.